### PR TITLE
Enable experimental features in frontend tests

### PR DESCRIPTION
**Pull request by DedSec256**

### CODE REVIEW

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpModel.kt [hunk 1]
```diff
@@ -15,6 +15,10 @@ object RdFSharpModel : Ext(SolutionModel.Solution) {
         +"Core"
     }
 
+    private val features = enum("RdExperimentalFeatures") {
```
Comment (auduchinok) [inline rider-fsharp/protocol/src/kotlin/model/RdFSharpModel.kt:18]: Should likely contain `FSharp`.

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpModel.kt [hunk 2]
```diff
@@ -15,6 +15,10 @@ object RdFSharpModel : Ext(SolutionModel.Solution) {
...
+        +"AllowFormatter"
```
Comment (auduchinok) [inline rider-fsharp/protocol/src/kotlin/model/RdFSharpModel.kt:19]: `Allow` seems redundant, i.e. the formatter is the feature.

File: ReSharper.FSharp/test/src/FSharp.Tests.Host/FSharpTestHost.fs [hunk 3]
```diff
@@ -41,12 +44,26 @@ type FSharpTestHost
...
             options.OtherOptions
             |> Array.choose (fun o -> if o.StartsWith("-r:") then Some (o.Substring("-r:".Length)) else None)
-            |> Array.map (fun s -> FileSystemPath.TryParse(s))
+            |> Array.map (FileSystemPath.TryParse)
             |> Array.filter (fun p -> not p.IsEmpty && directory.IsPrefixOf(p))
             |> Array.map (fun p -> p.Name)
             |> List)
         |> Option.defaultWith (fun _ -> List())
 
+    let setFeatures (rdFeatures: RdSetFeatures) =
```
Comment (auduchinok) [inline ReSharper.FSharp/test/src/FSharp.Tests.Host/FSharpTestHost.fs:53]: The name doesn't imply enabling/disabling at all, could we have a better one? please?

File: ReSharper.FSharp/test/src/FSharp.Tests.Host/FSharpTestHost.fs [hunk 4]
```diff
@@ -41,12 +44,26 @@ type FSharpTestHost
...
+        for feature in rdFeatures.Features do
+            match experimentalFeatures.TryGetValue feature with
+            | null -> ()
+            | value -> value.Dispose()
+
+            if rdFeatures.Enable then
+                experimentalFeatures.[feature] <-
+                    match feature with
+                    | RdExperimentalFeatures.AllowFormatter -> FSharpRegistryUtil.AllowFormatterCookie.Create()
```
Comment (auduchinok) [inline ReSharper.FSharp/test/src/FSharp.Tests.Host/FSharpTestHost.fs:62]: Could it be done in a more generic way, please? If we're to use it for other features, it'll be better not to duplicate it here for each one.

File: ReSharper.FSharp/test/src/FSharp.Tests.Host/FSharpTestHost.fs [hunk 5]
```diff
@@ -41,12 +44,26 @@ type FSharpTestHost
...
+                    | _ -> raise (InvalidOperationException("Unexpected experimental feature"))
+
+        JetBrains.Core.Unit.Instance
```
Comment (auduchinok) [inline ReSharper.FSharp/test/src/FSharp.Tests.Host/FSharpTestHost.fs:65]: Can this type be imported?

File: rider-fsharp/src/test/kotlin/extensions/TestExtensions.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package extensions
+
+import com.intellij.openapi.project.Project
+import com.jetbrains.rider.plugins.fsharp.RdExperimentalFeatures
+import com.jetbrains.rider.plugins.fsharp.RdSetFeatures
+import com.jetbrains.rider.plugins.fsharp.rdFSharpModel
+import com.jetbrains.rider.projectView.solution
+import com.jetbrains.rider.test.base.BaseTestWithSolutionBase
+
+fun BaseTestWithSolutionBase.withExperimentalFeatures(project: Project, features: Array<RdExperimentalFeatures>, function: () -> Unit) {
```
Comment (auduchinok) [inline rider-fsharp/src/test/kotlin/extensions/TestExtensions.kt:10]: Please add a test for it, it should check that the feature is actually enabled in the block and is disables afterwards.

**Comment by DedSec256** [comment]
We found another solution

**Pull request closed without merging**